File: LC1448-CountGoodNodesInBinaryTree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

int dfs(struct TreeNode* currRoot, int currMax) {
    if (!currRoot) {
        return 0;
    }

    // Check if current node is a "good" node
    // A node is good if its value >= maximum value on path from root
    int goodNodes = 0;
    if (currRoot->val >= currMax) {
        goodNodes = 1;
        currMax = currRoot->val;   // Update currMax children
    }

    // Recursively count good nodes in left and right subtrees
    goodNodes += dfs(currRoot->left, currMax);
    goodNodes += dfs(currRoot->right, currMax);
    return goodNodes;
}

int goodNodes(struct TreeNode* root){
    return dfs(root, root ? root->val : 0);
}



// Simple dfs solution - counting good node of left/right subtree recursively with dfs calls.
// Time Complexity: O(N) where N is the number of nodes in the binary tree
// Space Complexity: O(H) where H is the height of the binary tree (due to recursion stack)

File: LC1448-CountGoodNodesInBinaryTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        # DFS solution
        def dfs(curr_root: TreeNode, curr_max: int) -> int:
            if curr_root is None:
                return 0

            good_nodes = 0
            if curr_root.val >= curr_max:
                good_nodes = 1
                curr_max = curr_root.val

            good_nodes += dfs(curr_root.left, curr_max)
            good_nodes += dfs(curr_root.right, curr_max)
            return good_nodes

        return dfs(root, root.val if root is not None else 0)
